packages/core/upload/server/services/file.js

```diff
@@ -22,7 +23,34 @@ const deleteByIds = async (ids = []) => {
   return filesToDelete;
 };
 
+const signFileUrls = async (file) => {
+  const { provider } = strapi.plugins.upload;
+  const { provider: providerConfig } = strapi.config.get('plugin.upload');
+  const isPrivate = await provider.isPrivate();
+
+  // Check file provider and if provider is private
+  if (file.provider !== providerConfig || !isPrivate) {
+    return file;
+  }
+
+  const signUrl = async (file) => {
+    const signedUrl = await provider.getSignedUrl(file);
+    file.url = signedUrl.url;
+  };
+
+  const signedFile = cloneDeep(file);
+
+  // Sign each file format
+  await signUrl(signedFile);
+  if (file.formats) {
+    await mapAsync(Object.values(signedFile.formats), signUrl);
+  }
```

Comment:
Is this doing mutation? 😅

Reply:
a mutation on the cloned entity, do you think it's cleaner to not do it this way? 

I was trying to find other alternatives but I found this one easy to read. wdyt?

Reply:
Mutation are to avoid, we should never have some. It's really complicated to follow where things are being modified with mutation 🙁

Reply:
but don't you think in this case it's clean to follow?

I feel it's easy to follow when things are being modified if you are mutating an object you defined in the same function.
Otherwise  Iwould need a more complicated map iterating all fields.

And because of tyring to avoid mutations, would make the code harder to read.

Reply:
to me mutating `signedFile` is okay in this case. It is a clone of the file entity and is defined within the scope of this function so shouldn't have an impact elsewhere

Reply:
I understand what you say, and I would agree for the part where the mutation is contained in the function's body (and the function is relatively small and easy to read).
Le problème c'est que je suis tombé de nombreuses fois sur ce type de mutation où au final, la fonction a évolué avec le temps et tout est devenu illisible et incompréhensible, engendrant des bugs à se casser la tête pour trouver la source à cause d'une mutation.
La fonction `signUrl(file)` mute son argument `file`.
Pourquoi ne pas retourner l'url signée ?
```javascript
const signUrl = async (file) => {
  const signedUrl = await provider.getSignedUrl(file);
  return signedUrl.url;
};
// Sign each file format
signedFile.url = await signUrl(signedFile);
if (file.formats) {
  signedFile.formats = await mapAsync(Object.values(signedFile.formats), async (format) => {
    const url = await signUrl(format);
    return { ...format, url };
  });
}
```